pkg/daemon/certificate_writer.go

```diff
@@ -340,41 +339,40 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
 
 	klog.Infof("Certificate was synced from controllerconfig resourceVersion %s", controllerConfig.ObjectMeta.ResourceVersion)
 	if controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] == ctrlcommon.ServiceCARotateTrue && oldAnno != controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation] && cmErr == nil && kubeConfigDiff && !allCertsThere && !dontRestartKubelet {
-		logSystem("restarting kubelet due to server-ca rotation")
-		if err := runCmdSync("systemctl", "stop", "kubelet"); err != nil {
-			return err
-		}
-		err = os.Remove("/var/lib/kubelet/kubeconfig")
-		if err != nil {
-			return fmt.Errorf("could not remove kubelet's kubeconfig file: %v", err)
-		}
-		if err := runCmdSync("systemctl", "daemon-reload"); err != nil {
-			return err
-		}
-
-		if err := runCmdSync("systemctl", "restart", "kubelet"); err != nil {
-			return err
-		}
-
-		if err := wait.PollUntilContextTimeout(context.TODO(), 10*time.Second, 2*time.Minute, false, func(ctx context.Context) (bool, error) {
+		if len(currentKC.Clusters[0].Cluster.CertificateAuthorityData) > 0 {
+			logSystem("restarting kubelet due to server-ca rotation")
+			if err := runCmdSync("systemctl", "stop", "kubelet"); err != nil {
+				return err
+			}
 			f, err := os.ReadFile("/var/lib/kubelet/kubeconfig")
 			if err != nil && os.IsNotExist(err) {
 				klog.Warningf("Failed to get kubeconfig file: %v", err)
-				return false, nil
+				return err
 			} else if err != nil {
-				return false, fmt.Errorf("unexpected error reading kubeconfig file, %v", err)
+				return fmt.Errorf("unexpected error reading kubeconfig file, %v", err)
 			}
-			currentKC := clientcmdv1.Config{}
-			err = yaml.Unmarshal(f, &currentKC)
+			kubeletKC := clientcmdv1.Config{}
+			err = yaml.Unmarshal(f, &kubeletKC)
+			// set CA data to the one we just parsed above, the rest of the data should be preserved.
+			kubeletKC.Clusters[0].Cluster.CertificateAuthorityData = currentKC.Clusters[0].Cluster.CertificateAuthorityData
```

Comment:
This should be fine but I wonder if `Clusters` could ever contain more than 1 entry

Reply:
I never have seen that, clusters[0] is always the data we are looking for.